### PR TITLE
Rewrite auto-remove to always auto-remove all orphaned dependencies

### PR DESCRIPTION
**Pull request by GhostLyrics**
Let me know if this works for you or if I should try to rebuild it into a new parameter (though I’d honestly prefer not to do that).

### CODE REVIEW

**Comment by watercrossing** [comment]
This works perfectly on 14.04. Thanks!

**Comment by Burstaholic** [comment]
:+1:  This is badly needed!

**Comment by mvo5** [comment]
Hey, thanks for the branch and sorry for my slow reply. This branch looks great. I'm sitting on the fence a bit about adding a new option or not. Technically this is a behaviour change but given that a) the old feature apparently never worked reliable b) the wording is already talking about removing dependencies I think its fine to merge it. The thing I would like to see is some tests and extraction of the common code, I will look into this now.

**Comment by GhostLyrics** [comment]
I just cleaned up the code to remove all the pep8 warnings. Sorry for that, don't have python3 on my Mac. Please re-check the early-exit condition in case I was too eager to simplify.

Do you have a primer for starting with the tests?

**Comment by mvo5** [comment]
Hey Alexander! Looks like we may have a mid-air collision here :) I pushed what I had in mind for the cleanup here: https://github.com/mvo5/unattended-upgrades/tree/GhostLyrics-debian/jessie, its just extracting some common code. If you could merge that into your branch, that would rock. As for the tests, I have a look as the code is not that great when it comes to testability :/

**Comment by mvo5** [comment]
Oh and thanks for the pep8 cleanup!

**Comment by GhostLyrics** [comment]
@mvo5 not sure if that was by accident, but in GhostLyrics-debian/jessie you removed the possibility to test removal via `--dry-run` as far as I can see.

**Comment by a-detiste** [comment]
Hi,

I'd like to test this, but I don't have any auto-removable package on my systems at the moment and I don't know how to reliably recreate this condition.

I tried this
* `deborphan`
* pickup some package I don't really need
* `aptitude markauto $(package)` : this either remove package immediately or doesn't do anything
   I have to use the TUI to mark it auto
* then neither `apt-get autoremove` nor U-U detects this package as autoremovable

**Comment by a-detiste** [comment]
(where faking upgrades is a simple matter of pinning random things from sid/experimental)

**Comment by GhostLyrics** [comment]
I like to test in a VM with something along the lines of: `sudo apt-get install YOURPACKAGE (e.g. nodejs) && sudo apt-get remove YOURPACKAGE`

```
test@ubuntu-vm:~$ sudo apt-get install nodejs
[sudo] password for alex:
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following extra packages will be installed:
  libc-ares2 libv8-3.14.5
The following NEW packages will be installed:
  libc-ares2 libv8-3.14.5 nodejs
0 upgraded, 3 newly installed, 0 to remove and 17 not upgraded.
Need to get 1,912 kB of archives.
After this operation, 7,538 kB of additional disk space will be used.
Do you want to continue? [Y/n]
Get:1 http://de.archive.ubuntu.com/ubuntu/ trusty/main libc-ares2 amd64 1.10.0-2 [38.5 kB]
Get:2 http://de.archive.ubuntu.com/ubuntu/ trusty/universe libv8-3.14.5 amd64 3.14.5.8-5ubuntu2 [1,189 kB]
Get:3 http://de.archive.ubuntu.com/ubuntu/ trusty/universe nodejs amd64 0.10.25~dfsg2-2ubuntu1 [684 kB]
Fetched 1,912 kB in 1s (1,351 kB/s)
Selecting previously unselected package libc-ares2:amd64.
(Reading database ... 196292 files and directories currently installed.)
Preparing to unpack .../libc-ares2_1.10.0-2_amd64.deb ...
Unpacking libc-ares2:amd64 (1.10.0-2) ...
Selecting previously unselected package libv8-3.14.5.
Preparing to unpack .../libv8-3.14.5_3.14.5.8-5ubuntu2_amd64.deb ...
Unpacking libv8-3.14.5 (3.14.5.8-5ubuntu2) ...
Selecting previously unselected package nodejs.
Preparing to unpack .../nodejs_0.10.25~dfsg2-2ubuntu1_amd64.deb ...
Unpacking nodejs (0.10.25~dfsg2-2ubuntu1) ...
Processing triggers for doc-base (0.10.5) ...
Processing 1 added doc-base file...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Setting up libc-ares2:amd64 (1.10.0-2) ...
Setting up libv8-3.14.5 (3.14.5.8-5ubuntu2) ...
Setting up nodejs (0.10.25~dfsg2-2ubuntu1) ...
update-alternatives: using /usr/bin/nodejs to provide /usr/bin/js (js) in auto mode
Processing triggers for libc-bin (2.19-0ubuntu6.6) ...
alex@alex-desko:~$ apt-get remove nodejs
E: Could not open lock file /var/lib/dpkg/lock - open (13: Permission denied)
E: Unable to lock the administration directory (/var/lib/dpkg/), are you root?
alex@alex-desko:~$ sudo !!
sudo apt-get remove nodejs
Reading package lists... Done
Building dependency tree
Reading state information... Done
The following packages were automatically installed and are no longer required:
  libc-ares2 libv8-3.14.5
Use 'apt-get autoremove' to remove them.
The following packages will be REMOVED:
  nodejs
0 upgraded, 0 newly installed, 1 to remove and 17 not upgraded.
After this operation, 3,116 kB disk space will be freed.
Do you want to continue? [Y/n]
(Reading database ... 196393 files and directories currently installed.)
Removing nodejs (0.10.25~dfsg2-2ubuntu1) ...
Processing triggers for man-db (2.6.7.1-1ubuntu1) ...
Processing triggers for doc-base (0.10.5) ...
Processing 1 removed doc-base file...
```

The important part here is:
```
The following packages were automatically installed and are no longer required:
  libc-ares2 libv8-3.14.5
Use 'apt-get autoremove' to remove them.
```

**Comment by mvo5** [comment]
I added a test in https://github.com/mvo5/unattended-upgrades/commit/0ece9e44e91ff8178ac767027319124743e7deb1 that simulates the dependency removal.

**Comment by mvo5** [comment]
If you merge my changes I'm happy to merge your branch (this way its probably easiest). As for the dry-run option. Its fine to call cache_commit() with dry-run, it will not really do anything expect for printing the dpkg commands it would run. So unless I miss someting (quite possible :) my removal of the "if options.dry_run:" should not have done any harm. If I did miss something, please help me to understand what!

**Comment by mvo5** [comment]
Hey a-detiste , another way of testing this is: "sudo apt-get install 2vcard; sudo apt-mark auto 2vcard; sudo ./unattended-upgrades -v" (with auto-removal enabled). This should remove 2vcard again.

**Comment by mvo5** [comment]
Thanks again, I resolved the conflicts and merged into the debian/sid branch (which is the master branch for unattended-upgrades).

**Comment by a-detiste** [comment]
> # exit if there is nothing to do and nothing to report

With the recent changes, UU now sends empty mails when it has nothing to do.

**Comment by GhostLyrics** [comment]
@a-detiste do you happen to have missed the following?

```
// Set this value to "true" to get emails only on errors. Default
// is to always send a mail if Unattended-Upgrade::Mail is set
//Unattended-Upgrade::MailOnlyOnError "true";
```

File: unattended-upgrade [hunk 1]
```diff
@@ -1205,27 +1245,19 @@ def main(options, rootdir=""):
...
-        for pkgname in now_auto_removable - pkgs_auto_removable:
-            logging.debug("marking %s for remove" % pkgname)
-            cache[pkgname].mark_delete()
-        logging.info(_("Packages that are auto removed: '%s'") %
-                     " ".join(now_auto_removable - pkgs_auto_removable))
-
-    logging.debug("InstCount=%i DelCount=%i BrokenCount=%i" % (
-        cache._depcache.inst_count, cache._depcache.del_count,
-        cache._depcache.broken_count))
-
```
Comment (a-detiste) [inline unattended-upgrade]: @GhostLyrics here

**Comment by a-detiste** [comment]
Even with `Unattended-Upgrade::MailOnlyOnError "false";`, no mail is sent when UU has nothing to do.

**Comment by mvo5** [comment]
That might be my changes, I will see how/why I broke this.

**Comment by mvo5** [comment]
@a-detiste this https://github.com/mvo5/unattended-upgrades/commit/782f9fb7207d6fcf7a7dcd8ee281de59e74410c4 should fix the mail issue you see, the problem was that the commit() of a empty change in auto-remove triggers a mail.

**Comment by mvo5** [comment]
And thanks for the testing, please let us know if there is more :)

**Comment by a-detiste** [comment]
I got this
```
python3 test_remove_unused.py
No such file or directory: '/home/pi/unattended-upgrades/test/root.unused-deps/var/lib/dpkg/updates
```

**Comment by mvo5** [comment]
Thanks! Silly me, I keep forgeting that git does not track dirs, I added a
empty placeholder now.

On Tue, May 26, 2015 at 2:28 PM, Alexandre Detiste <notifications@github.com

**Comment by a-detiste** [comment]
I'm still getting the "empty" mails.

**Comment by a-detiste** [comment]
Now I've got this ;-)

```
python3 test_remove_unused.py                                                                   
An error occurred: '404  Not Found [IP : 91.189.91.24 80]'                                                                                  
The URI 'http://archive.ubuntu.com/ubuntu/test-package_2.0_all.deb' failed to download, aborting
```

**Comment by mvo5** [comment]
Thanks! The test should be fixed now and I also changed the code a bit so
that the removal of unused dependencies is only done if there is a upgrade.
The reason for the empty mails was that the previous check if there is
anythig to do was changed and u-u always acts and sends mail when the
option "remove-unused-dependencies" is set.

On Tue, May 26, 2015 at 3:13 PM, Alexandre Detiste <notifications@github.com

**Pull request merged**